### ISSUE TITLE
Feture request: parallel download

### ISSUE THREAD

**Issue by MauroMombelli** [trizen/trizen#127]
An idea to implement parallel download:
- run pacman -Sup (no root required) to get the list of packet requiring download
- download the pagages witha downlaod manager that support multple source, like aria2
- move the downloaded files into pacman cache (require permission)
- run pacman -Syu (require permission)

**Comment by trizen** [comment]
Take a look at [powerpill](https://wiki.archlinux.org/index.php/Powerpill).

Does changing the `pacman_command` to use powerpill works? (I can't test right now)

If not, then maybe adding a new config-option to use powerpill only for sync operations should be fine.

**Comment by MauroMombelli** [comment]
SHORT STORY: if you can add an option to call an external script for download operation, that would be perfect!
Some suggestion the informatin needed to simplify the download script (would be nice to have them from stdin/out, all in one call):
1. list of mirror
2. desitnation directory
3. for each package:
 - architecture (can be x86_64 or generic)
 - pkg name
 - pkg name
 - pkg version (some pkg name make hard to slipt name and version)
 - repository
4. a way to tell trizen the download has finish, and proceed to install (possibly without killing the script)

FOR AUR:
1. download sources (URLs)
2. download directory
3. a way to tell trizen the download has finish, and proceed to install

for AUR you will have to manage to wait for dependecy, I guess, but that can also be extended to pacman, so we will ave parallel download AND installation

LONG STORY:
yes i guess it would work but then i would use powerpill, but that come with an extremly big list of (AUR) dependency, and the guy developing it still havent make the repo public, so i would like to avoid it.

I develop this bash script (https://gist.github.com/MauroMombelli/04b24fa0644a4870869099276b86f2d4) that will run te parallel download (by default aria2c uses 5 server, 1 connection for server, and i randomize the list of mirror so every time you use different, if you decomment more than 5 in mirrorlist), the only dipendency are "aria2" and "sudo".
Sudo dependency is only neeed to save the packages in the pacman cache folder (by default "/var/pacman/cache", but is read from pacman.conf), but if a temporary cache or some other magic is used, that can be dropped. (but i like to keep it "default", so other program play nice, like "downgrader").

from my test, form a capper 1MB/s download, i can finally saturate my bandwith (6MB/s)
![image](https://user-images.githubusercontent.com/6372561/37654133-5ceb7cde-2c41-11e8-82ec-f376fb41e21a.png)

this will only work for "pacman" pagages, while the same could be applied also for the AUR packages, parallel download.

Finally im using aria2c (FTP/HTTP/HTTPS), but looking at powerpill many mirror support "rsync", that should give a great boost on many small files, basically all the core and most of the oter repos too.

why an external script?
Because you dont have to worry about HOW we download stuff.

**Comment by AladW** [comment]
What a long rant just to describe something you can already do...

**Comment by MauroMombelli** [comment]
no, you can *partially* do it. Especially the parallel download of AUR artifact is something nodoby does, even powerpill. Not to talk about the parallel download + installation.

But please, if I'm wrong instead of a passive-agressive comment that add nothing to the discussion, leave a link.

**Comment by Morganamilo** [comment]
parallel installation is just not a thing that's feasible. There's a build of Yay with parallel AUR downloading and even though it does work and gives a speedup I still don't know if it's a good idea because of the messed up printing and error handling.

**Comment by AladW** [comment]
Parallel downloading of tar archives might have some use, though mostly not since the AUR snapshots are very small and the total download time pales compared to the actual build. Trying to do the same with `git` is in my experience just asking for trouble (e.g. I had all kinds of strange errors when parallel cloning over ssh).

I have no idea what parallel "installation" is supposed to mean, if you mean parallel builds you can either do the sensible things and change your makeflags or add an [absurd amount of complexity](https://github.com/jrmarino/synth) for minimal gain.

**Comment by Morganamilo** [comment]
When I was talking about parallel AUR downloads I was on about downloading the sources=() files, not the PKGBUILDS themselves. It give me a 30% speedup in my narrow testing. I still wouldn't call it worth it though, there's a bunch of negatives.

**Comment by AladW** [comment]
That seems outside the scope of an AUR helper to me. It might fit into `makepkg` itself or user configuration.

**Comment by Morganamilo** [comment]
The problem with makepkg is it only handles one pkgbuild at a time. Plus the way it was done was by running `makepkg --nobuild` on all pkgbuilds at once. So each source in a pkgbuild was built sequentially but all to be installed packages were downloaded in paralleled.

The biggest problem was all the output to stdout and secondly how to handle errors.

Just want to make note of it in case anybody wants to continue the idea and come up with a better solution for it.

**Comment by MauroMombelli** [comment]
parallel installation of multiple package is not possible, but installation while downloading is possible.
Parallel download of AUR pakgages has the same issue of mixing download of "normal" packages, and powerpill is a clear demonstration it can be done, so calling for error managment and UX issue is not a valid point.

>Trying to do the same with git

you are right, as the git clone/pull can't be optimized right now, and even if it could as the server is tipically one it would not be very etic, as would be avoiding speed limit (while with pacman we use different mirror so no harm is really done).
But the idea here is to start multiple packet download in parallel.

my tipical case is checking out linux-mainline, that is pretty big and slow, and block other smaller packet from updating. Same issue with some big games like 0AD that download from their own server.
Also in that particular case the *build* phase could be parallelized, but i digress and should be a different feature with its own can of worms.

as @morganamilo say, this apply to source(), not to the PKBGBUILDs

For me the spedup on regular pakages is from 1MB/s to 5MB/s (line saturation), this is 500%.

>That seems outside the scope of an AUR helper to me. It might fit into makepkg itself or user configuration.

quite interesting. first of all the fact to divide the download process from the management process is exaclty what i proposed in my long exmplanation, actually with a raw desciption on how i would implement the communications.
I don't see how this could be done by makepkg

**Comment by AladW** [comment]
>installation while downloading is possible.

You seem to underestimate the complications involved. For example, you could only proceed with the following package if it doesn't depend on any previous ones. Trizen has a strictly linear workflow where it "sees one package at a time", and even if you solved #89 you'd need additional complicated mechanisms to make it work for little gain in the end.

 >I don't see how this could be done by makepkg

makepkg downloads the actual source files in the first place... so you set DLAGENT to e.g. `aria2` and done.

https://wiki.archlinux.org/index.php/aria2#Using_Aria2_with_makepkg

**Comment by MauroMombelli** [comment]
>makepkg downloads the actual source files in the first place... so you set DLAGENT to e.g. aria2 and done.

but this part can not be parallelized, as mostly come from only one server, is the download of mutiple AUR packages that can be parallelized.

>Trizen has a strictly linear workflow where it "sees one package at a time"

Please don't mix the issue of building, it add confusion and is out of scope. We talk about "parallel" installation of nomal packages, and if you say "linear", it means you are waiting for a specific order, then instead of starting 10 download with 4 mirrot, i can start he next 4 download with 10 mirror, thus priorizing the queue. I won't go under 4 parallel dowload as in my experience there is were the real "bandwidth filler" is done, skipping the "setup of download" phase for smaller pkg

```>>> 19:11:45 | Downloading git, 1/10 download
>>> 19:11:45 | Downloading lib32-tdb, 1/10 download
>>> 19:11:45 | Downloading libmagick6, 2/10 download
>>> 19:11:45 | Downloading linux-lts, 3/10 download
>>> 19:11:46 | Downloading telegram-desktop, 2/10 download
>>> 19:11:46 | Downloading tzdata, 2/10 download```

**Comment by AladW** [comment]
>Please don't mix the issue of building, it add confusion and is out of scope. 

It's not about building. You said "downloading while installation of packages" and if you want to install some package that depends on another one that hasn't even finished downloading there's no point.

Anyway there's a clear language barrier here and the more we argue the less I understand what you're trying to do. So I'll just stick to https://github.com/trizen/trizen/issues/127#issuecomment-376250020 and suggest you look at GNU parallel.

**Comment by MauroMombelli** [comment]
ok I sense confusion.

>downloading while installation of packages

yes, but this is for non-aur. For aur normally forst you have to compile, and that is a sincronous operation, as #89 

>suggest you look at GNU parallel

i don't see how this could help, you want to split the compilation on multiple machine? but this issue is NOT about building, plus it seems to me a vastly complicated stuff. A dedicated build server for aur stuff would then be easier.

so to clarify; this issue is about implementi an interface to communicate with an external program/service that manage the download, and can handle parallel download of packages and "source=()".

There are a lot of things trizen can handle like giving the list of packages based on the priority, mixing aur and normal packages, and who knows what can be done; but that is out of scope for this request.

The native implementation/integration of a downalod system would open the door to many possibility, maybe even requiring full restructure of the code.

**Comment by AladW** [comment]
>you want to split the compilation on multiple machine

???

From https://github.com/trizen/trizen/issues/127#issuecomment-376246234

>The biggest problem was all the output to stdout and secondly how to handle errors.

To which I replied:

>Well, if you find that compelling you should look at how GNU Parallel does it. Errors can be handled in various ways (e.g. exit directly when any of the processes fails) and the output is buffered nicely.

**Comment by joeky888** [comment]
> Run pacman -Sup (no root required) to get the list of packet requiring download

Thanks for this!

I wrote a script called apt-fast on Ubuntu/Debian and just ported it to Archlinux, the only dependency is aria2.

Usage on Ubuntu/Debian: `apt-fast install vlc -y`
Usage on Archlinux: `apt-fast -S vlc`


Save it to somewhere like `/usr/bin/apt-fast`

```sh
#!/bin/bash

[ "`whoami`" = root ] || exec sudo "$0" "$@"

APT=$([ $(command -v apt) ] && echo "apt" || echo "apt-get")
InstallNoAsk="install -y"

if [ $(command -v pacman) ]; then
  APT="pacman"
  InstallNoAsk="-S --noconfirm --needed"
fi

if echo "$@" | grep -q "upgrade\|install\|dist-upgrade\|full-upgrade\|build-dep\|-S"; then
  cd /var/cache/apt/archives/ 2>/dev/null;
  cd /var/cache/pacman/pkg/ 2>/dev/null;

  DOWNLOADARGS='-c -s16 -k1M -x16'
  DNS=""

  if ping -c 1 -W 2 8.8.8.8 2>/dev/null | grep -q -i "ttl"; then
    DNS="--async-dns-server=8.8.8.8"
  fi

  # Donload unfinished jobs
  [ $(command -v aria2c) ] || $APT $InstallNoAsk aria2
  [[ -f /var/cache/apt/archives/apt-fast.list ]] && aria2c $DOWNLOADARGS $DNS -i apt-fast.list
  [[ -f /var/cache/pacman/pkg/apt-fast.list ]] && aria2c $DOWNLOADARGS $DNS -i apt-fast.list

  # Create new jobs
  if [ $(command -v apt-get) ]; then
    apt-get -y --print-uris $@ | grep -E -o -e "(ht|f)tp[s]?://[^\']+" | tee apt-fast.list && aria2c -c $DOWNLOADARGS $DNS -i apt-fast.list
  elif [ $(command -v pacman) ]; then
    pacman -Syup | grep -E -o -e "(ht|f)tp[s]?://[^\']+" | tee apt-fast.list && aria2c -c $DOWNLOADARGS $DNS -i apt-fast.list
    pacman -Sup $(echo "${@:1}" | sed "s/-[a-zA-Z]\+//g") | grep -E -o -e "(ht|f)tp[s]?://[^\']+" | tee apt-fast.list && aria2c -c $DOWNLOADARGS $DNS -i apt-fast.list
  fi

elif echo "$@" | grep -q "update"; then
  $APT -o 'Acquire::Queue-mode=access; APT::Acquire=Retries 3;' update
fi

echo $APT $@
$APT $@ && rm -f /var/cache/apt/archives/apt-fast.list && rm -f /var/cache/pacman/pkg/apt-fast.list
```

**Comment by MauroMombelli** [comment]
In tge end i also create a wrapper script, it will sue aria2 to download
all the updates (not AUR) and place them in pacman's cache folder. See
https://github.com/lestofante/PacmanParallelizer

**Comment by MauroMombelli** [comment]
Closing as implemented in Pacman v6, see http://allanmcrae.com/2020/12/pacman-6-0-0alpha1/
